R/valley.R

```diff
@@ -0,0 +1,273 @@
+#' 
+#' @return Object reprojected to specified CRS
+#' @export
+reproject <- function(x, crs, ...){
+  if (inherits(x, "SpatRaster")){
+    return(terra::project(x, crs, ...)
+  } else if (inherits(x, c("bbox", "sfc", "sf"))) {
+    return(sf::st_transform(x, crs, ...))
+  } else {
+    stop(sprintf("Cannot reproject object type: %s", class(x)))
+  }  
+}
+
+#' spatially smooth dem by (window) filtering
+#' 
+#' @param dem raster data of dem 
+#' @param method filtering function to be used, e.g. "median". As accepted by focal
+#' @param window size of filter
+#' 
+#' @return filtered dem
+#' @export
+filter_dem <- function{dem, method="median", window=5}{
+    dem_filtered <- focal(dem, w=window, fun=method)
+    names(dem_filtered) <- "dem_filtered"
+    return(dem_filtered)
+}
+
+#' Derive slope as percentage from dem
+#' This makes use of the terrain function of the terra package 
+#' 
+#' @param dem raster data of dem 
+#' 
+#' @return raster of derived slope over dem extent
+#' @export
+get_slope <- function(dem){
+    slope_radians <- terrain(dem, v = "slope", unit = "radians")
+    slope <- tan(slope_radians)
+}
+
+#' Mask slope raster, setting the slope to zero for the pixels overlapping 
+#' the river area.
+#' 
+#' @param slope raster data of slope
+#' @param vector/polygon data of river
+#' @param lthresh lower numerival threshold to consider slope non-zero 
+#' 
+#' @return updated slope raster
+#' @export
+mask_slope <- function(slope, river, lthresh=1.e-3, target = 0){
+    slope_masked <- mask(
+    slope,
+    ifel(slope <= lthresh, NA, 1),
+    updatevalue = lthresh)
+
+    slope_masked <- mask(
+    slope_masked,
+    river,
+    inverse = TRUE,
+    updatevalue = target,
+    touches = TRUE)
+}
+
+#' Derive cost distance function from masked slope
+#' 
+#' @param slope_masked raster of masked slope data
+#' @param target value for codt distance calculation
+#' 
+#' @return raster of cost distance
+#' @export
+get_cost_distance <- function(slope, river, target = 0){
+  slope_masked <- mask_slope(slope, river, target = target)
+  cd <- costDist(slope_masked, target = target)
+  names(cd) <- "cost_distance"
+  return(cd)
+}
+
+#' Mask out river regions incl. a buffer in cost distance raster data
+#' 
+#' @param cd cost distance raster
+#' @param river vector/polygon
+#' @param BUFFER size of buffer around river polygon to additionally mask
+#' 
+#' @return cd raster with river+BUFFER pixels masked
+#' @export
+mask_cost_distance <- function(cd, river, BUFFER=2000){
+    BUFFER_REGION <- BUFFER  # m
+    river_buffer <- st_buffer(river, BUFFER_REGION)
+    cd_masked <- mask(
+        cd,
+        river_buffer,
+        updatevalue = NA,
+        touches = TRUE)
+}
+
+#' Get characteristic value of distribution of cost distance
+#'
+#' @param cd cost distance raster data
+#' @param method function used to derive caracteristic value (mean)
+#' 
+#' @return characteristic value of cd raster
+#' 
```

Comment:
I see that some helper functions (like `mask_slope()`) are exported and some are not. In principle, we only need to export the ones that we want to be visible to the user.